File: liveobjects_8/Quilt/Transmitter/Transmitter.cs
/*

Copyright (c) 2004-2009 Qi Huang. All rights reserved.

Redistribution and use in source and binary forms,
with or without modification, are permitted provided that the following conditions
are met:

1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above
   copyright notice, this list of conditions and the following
   disclaimer in the documentation and/or other materials provided
   with the distribution.

THIS SOFTWARE IS PROVIDED "AS IS" BY THE ABOVE COPYRIGHT HOLDER(S)
AND ALL OTHER CONTRIBUTORS AND ANY EXPRESS OR IMPLIED WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDER(S) OR ANY OTHER
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

*/

#define VERBOSE

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using QS.Fx.Value;
using QS.Fx.Serialization;

/// Transmitter is the network module residing in Quilt taking care 
/// of network communication.
/// 
/// It is a client of EUIDTransport, maintaining all channels inside,
/// while helping sending out message and dispatching received messages
/// to upper modules
/// 
namespace Quilt.Transmitter
{
    public class Transmitter :
        QS._qss_x_.Properties_.Component_.Base_,
        QS.Fx.Interface.Classes.ITransportClient<EUIDAddress, TransmitterMsg>
    {
        #region Delegates

        // Delegate function to handle received message
        public delegate void UpperMessageHandler(EUIDAddress remote_euid, ISerializable message);

        #endregion

        #region Constructor

        public Transmitter(
            QS.Fx.Object.IContext _mycontext,
            QS.Fx.Object.IReference<QS.Fx.Object.Classes.ITransport<EUIDAddress, TransmitterMsg>>
            _transport_object_reference,
            UpperMessageHandler 
            _upper_message_handler)
            : base(_mycontext, true)
        {
#if VERBOSE
            if (this._logger != null)
                this._logger.Log("Quilt.Transmitter.Transmitter constructor");
#endif

            this._mycontext = _mycontext;

            // Connect EUIDTransport
            this._transport_endpt = this._mycontext.DualInterface<
                QS.Fx.Interface.Classes.ITransport<EUIDAddress, TransmitterMsg>,
                QS.Fx.Interface.Classes.ITransportClient<EUIDAddress, TransmitterMsg>>(this);

            this._transport_endpt.OnConnected +=
                new QS.Fx.Base.Callback(
                    delegate { this._Enqueue(new QS._qss_x_.Properties_.Base_.Event_(new QS._qss_x_.Properties_.Base_.EventCallback_(this._Connect))); });
            this._transport_endpt.OnDisconnect +=
                new QS.Fx.Base.Callback(
                    delegate { this._Enqueue(new QS._qss_x_.Properties_.Base_.Event_(new QS._qss_x_.Properties_.Base_.EventCallback_(this._Disconnect))); });

            this._transport_conn = this._transport_endpt.Connect(_transport_object_reference.Dereference(_mycontext).Transport);

            // Initialize containers
            this._channel_dict = new Dictionary<string, ChannelHandler>();
            this._waiting_dict = new Dictionary<string, Queue<ISerializable>>();
            this._outgoing_queue = new Queue<Tuple_<EUIDAddress, ISerializable>>();
            this._incoming_queue = new Queue<Tuple_<EUIDAddress, ISerializable>>();

            this._message_handler = MessageHandler;
            this._channel_disconnect_handler = ChannelBreakHandler;

            this._upper_message_handler = _upper_message_handler;
        }

        #endregion

        #region Fields

        [QS.Fx.Base.Inspectable]
        private QS.Fx.Object.IContext _mycontext;

        [QS.Fx.Base.Inspectable]
        private QS.Fx.Endpoint.Internal.IDualInterface<
            QS.Fx.Interface.Classes.ITransport<EUIDAddress, TransmitterMsg>,
            QS.Fx.Interface.Classes.ITransportClient<EUIDAddress, TransmitterMsg>> _transport_endpt;

        private QS.Fx.Endpoint.IConnection _transport_conn;

        private EUIDAddress _self_euid;

        private Dictionary<string, ChannelHandler> _channel_dict;
        private Dictionary<string, Queue<ISerializable>> _waiting_dict;
        private Queue<Tuple_<EUIDAddress, ISerializable>> _outgoing_queue;
        private Queue<Tuple_<EUIDAddress, ISerializable>> _incoming_queue;

        private ChannelHandler.MessageHandler _message_handler;
        private ChannelHandler.ChannelBreakHandler _channel_disconnect_handler;

        private UpperMessageHandler _upper_message_handler;

        #endregion

#if VERBOSE

        // Debug code for logging transmitter message flow
        private System.IO.StreamWriter _peer_log = null;
        public void SetLogTarget(System.IO.StreamWriter _peer_log)
        {
            this._peer_log = _peer_log;
        }
#endif

        // @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

        #region _Connect for EUIDTransport

        private void _Connect(QS._qss_x_.Properties_.Base_.IEvent_ _event)
        {
#if VERBOSE
            if (this._logger != null)
                this._logger.Log("Quilt.Transmitter.Transmitter EUIDTransport Endpoint Connect");
#endif

        }

        #endregion

        #region _Disconnect for EUIDTransport

        private void _Disconnect(QS._qss_x_.Properties_.Base_.IEvent_ _event)
        {
#if VERBOSE
            if (this._logger != null)
                this._logger.Log("Quilt.Transmitter.Transmitter EUIDTransport Endpoint Disconnect");
#endif
            // Clear channel dictionary
            this._channel_dict.Clear();
        }

        #endregion

        // @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

        #region ITransportClient<EUIDAddress,TransmitterMsg> Members

        void QS.Fx.Interface.Classes.ITransportClient<EUIDAddress, TransmitterMsg>.Address(EUIDAddress address)
        {
#if VERBOSE
            if (this._logger != null)
                this._logger.Log("Quilt.Transmitter.Transmitter.ITransportClient.Address" + address.String);
#endif
            this._self_euid = address;
            MessageHandler(_self_euid, null);
        }

        void QS.Fx.Interface.Classes.ITransportClient<EUIDAddress, TransmitterMsg>.Connected(EUIDAddress address, QS.Fx.Object.IReference<QS.Fx.Object.Classes.ICommunicationChannel<TransmitterMsg>> channel)
        {
          
            try
            {
                if (!_channel_dict.ContainsKey(address.String))
                {
                    ChannelHandler channel_handler = new ChannelHandler(_mycontext, address, channel, _message_handler, _channel_disconnect_handler);
                    _channel_dict.Add(address.String, channel_handler);

#if VERBOSE
                    //if (this._logger != null)
                    //    this._logger.Log("Quilt.Transmitter.Transmitter.ITransportClient.Connected new " + address.String);
#endif  

                    if (_waiting_dict.ContainsKey(address.String))
                    {
                        // Need to clear the waiting messages
                        this._Enqueue(
                            new QS._qss_x_.Properties_.Base_.Event_<EUIDAddress>(
                            new QS._qss_x_.Properties_.Base_.EventCallback_(this.ClearMessage), address));
                    }
                }
            }
            catch (Exception e)
            {
                throw new Exception("Quilt.Transmitter. ITransportClient.Connected function, dictionary add exception: " + e.Message);
            }
        }

        #endregion

        // @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

        #region Message Process

        private void MessageProcess(QS._qss_x_.Properties_.Base_.IEvent_ _event)
        {
            // Processing messages
            while (_incoming_queue.Count > 0)
            {
                Tuple_<EUIDAddress, ISerializable> tuple = _incoming_queue.Dequeue();

                // Call Upper layer process functions
                _upper_message_handler(tuple.x, tuple.y);
            }
        }

        #endregion

        // @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

        #region MessageHandler and ChannelBreakHandler for ChannelHandler

        // These two functions would be called by Enqueueed messages

        private void MessageHandler(EUIDAddress remote_euid, TransmitterMsg message)
        {
            bool isProcessing = _incoming_queue.Count > 0;

            if (message == null)
            {
                _incoming_queue.Enqueue(new Tuple_<EUIDAddress, ISerializable>(remote_euid, null, 0));
            }
            else if (remote_euid.IsFull())
            {
                _incoming_queue.Enqueue(new Tuple_<EUIDAddress, ISerializable>(remote_euid, message.Message, 0));
            }
            else
            {

                EUIDAddress new_remote_euid = ((IEUIDable)message).GetEUID();

#if VERBOSE
                //if (this._logger != null)
                //    this._logger.Log("Quilt.Transmitter.Transmitter.Update EUID " + remote_euid.String + " to " + new_remote_euid.String);
#endif

                ///
                /// If the message comes from a incoming connection
                /// need to fix the remote EUID
                /// 
                if (new_remote_euid.IsFull())
                {
                    // Change the key of channel_dict
                    ChannelHandler new_channel;
                    if (_channel_dict.TryGetValue(remote_euid.String, out new_channel))
                    {
                        bool isReplacable = true;

                        ChannelHandler duplicated_channel;
                        // Check if there a duplicated outgoing connection to remote host
                        if (_channel_dict.TryGetValue(new_remote_euid.String, out duplicated_channel))
                        {
                            // Keep the connection starting from a smaller EUID, just for coherent selection
                            if (((IComparable<EUIDAddress>)_self_euid).CompareTo(new_remote_euid) <= 0)
                            {
                                isReplacable = false;
                            }
                            else
                            {
                                // Remove the duplicated channel
                                duplicated_channel.Disconnect();
                            }
                        }

                        if (isReplacable)
                        {
                            // Set the ChannelHandler
                            new_channel.RemoteEUID = new_remote_euid;

                            // Set the _channel_dict
                            _channel_dict[new_remote_euid.String] = new_channel;
                            _channel_dict.Remove(remote_euid.String);
                        }
                        else
                        {
                            // Keep the old one
                            new_channel.Disconnect();
                        }
                    }
                    _incoming_queue.Enqueue(new Tuple_<EUIDAddress, ISerializable>(new_remote_euid, message.Message, 0));
                }
            }
            

            if (!isProcessing)
            {
                this._Enqueue(
                    new QS._qss_x_.Properties_.Base_.Event_(
                    new QS._qss_x_.Properties_.Base_.EventCallback_(this.MessageProcess)));

            }
        }

        private void ChannelBreakHandler(EUIDAddress remote_euid)
        {
#if VERBOSE

            if (this._logger != null)
                this._logger.Log("Component_.Transmitter.ChannelBreakHandler Channel removed " + remote_euid.String);

#endif
            // Remove the channel from dictionary
            _channel_dict.Remove(remote_euid.String);
        }

        #endregion

        // @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

        #region SendMessage API

        public void SendMessage(EUIDAddress remote_euid, ISerializable message)
        {
            bool isSending = this._outgoing_queue.Count > 0;
            _outgoing_queue.Enqueue(new Tuple_<EUIDAddress, ISerializable>(remote_euid, message, 0));

            if (!isSending)
            {
                this._Enqueue(
                    new QS._qss_x_.Properties_.Base_.Event_(
                    new QS._qss_x_.Properties_.Base_.EventCallback_(this.SendMessage)));

            }
        }

        #endregion

        #region SendMessage, called by _Enqueue

        private void SendMessage(QS._qss_x_.Properties_.Base_.IEvent_ _event)
        {
            while (_outgoing_queue.Count > 0)
            {
                Tuple_<EUIDAddress, ISerializable> tuple = _outgoing_queue.Dequeue();
                EUIDAddress remote_euid = tuple.x;
                ISerializable message = tuple.y;

                ChannelHandler channel_handler;
                if (_channel_dict.TryGetValue(remote_euid.String, out channel_handler))
                {
#if VERBOSE
                    //if (this._logger != null)
                        //this._logger.Log(DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond + " Quilt.Transmitter.Transmitter.SendMessage begin");

                    if (_peer_log != null)
                    {
                        double cur = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;
                        _peer_log.WriteLine(cur + "\tTransmitter\t" + message.SerializableInfo.Size + "\t" + message.SerializableInfo.ClassID + "\t" + "\tPushed To Channel");
                        _peer_log.Flush();
                    }
#endif

                    // Wrap message with self EUID
                    TransmitterMsg tran_msg = new TransmitterMsg(this._self_euid, message);
                    channel_handler.Send(tran_msg);

#if VERBOSE
                    //if (this._logger != null)
                    //    this._logger.Log(DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond + " Quilt.Transmitter.Transmitter.SendMessage end");
#endif
                }
                else
                {
#if VERBOSE
                    //if (this._logger != null)
                    //    this._logger.Log("Quilt.Transmitter.Transmitter.SendMessage new channel for " + remote_euid.String);
#endif
                    // Set up waiting queue
                    Queue<ISerializable> waiting_queue;
                    if (!_waiting_dict.TryGetValue(remote_euid.String, out waiting_queue))
                    {
                        waiting_queue = new Queue<ISerializable>();
                        _waiting_dict.Add(remote_euid.String, waiting_queue);

                        // Connect remote end-host
                        _transport_endpt.Interface.Connect(remote_euid);
                    }
                    waiting_queue.Enqueue(message);
                   
                }
            }
        }

        #endregion

        #region Clear wating messages, called by _Enqueue

        private void ClearMessage(QS._qss_x_.Properties_.Base_.IEvent_ _event)
        {
            QS._qss_x_.Properties_.Base_.IEvent_<EUIDAddress> _event_ = (QS._qss_x_.Properties_.Base_.IEvent_<EUIDAddress>)_event;
            EUIDAddress remote_euid = _event_._Object;
            Queue<ISerializable> waiting_queue = _waiting_dict[remote_euid.String];
            ChannelHandler channel_handler = _channel_dict[remote_euid.String];

            while (waiting_queue.Count > 0)
            {
                // Wrap message with self EUID
                ISerializable message = waiting_queue.Dequeue();
                channel_handler.Send(new TransmitterMsg(_self_euid, message));

#if VERBOSE
                //if (this._logger != null)
                //this._logger.Log(DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond + " Quilt.Transmitter.Transmitter.SendMessage begin");

                if (_peer_log != null)
                {
                    double cur = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;
                    _peer_log.WriteLine(cur + "\tTransmitter\t" + message.SerializableInfo.Size + "\t" + message.SerializableInfo.ClassID + "\t" + "\tPushed To Channel");
                    _peer_log.Flush();
                }
#endif
            }

            _waiting_dict.Remove(remote_euid.String);
        }

        #endregion
    }
}

File: liveobjects_5/_core_c_/Core/Transmitter.cpp
/*

Copyright (c) 2004-2009 Krzysztof Ostrowski. All rights reserved.

Redistribution and use in source and binary forms,
with or without modification, are permitted provided that the following conditions
are met:

1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above
   copyright notice, this list of conditions and the following
   disclaimer in the documentation and/or other materials provided
   with the distribution.

THIS SOFTWARE IS PROVIDED "AS IS" BY THE ABOVE COPYRIGHT HOLDER(S)
AND ALL OTHER CONTRIBUTORS AND ANY EXPRESS OR IMPLIED WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDER(S) OR ANY OTHER
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

*/

#include "stdafx.h"
#include <winsock2.h>

#include "Transmitter.h"
#include "Overlapped.h"

using namespace System;
using namespace System::Threading;
using namespace System::Runtime::InteropServices;
using namespace System::Net;
using namespace System::Net::Sockets;
using namespace System::Collections::Generic;

// #define DEBUG_DisplayOutgoingPackets 

namespace QS
{
	namespace _core_c_
	{
		namespace Core
		{
			#pragma region Transmit

			void Transmitter::Transmit(QS::Fx::Network::AsynchronousSend request, int seqno)
			{
#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt1 = clock->Time;
#endif

				MyOverlapped* overlapped = (MyOverlapped *) poverlapped;
				ZeroMemory(&overlapped->overlapped, sizeof(WSAOVERLAPPED));

				int nbuffers = request.Data.Segments->Count;

#if defined(DEBUG_TagOutgoingStreams)
				nbuffers++;
#endif

				if (overlapped->nsenderbuffers < nbuffers)
				{
					delete overlapped->senderbuffers;

					while (overlapped->nsenderbuffers < nbuffers)
						overlapped->nsenderbuffers = 2 * overlapped->nsenderbuffers;
					overlapped->senderbuffers = new WSABUF[overlapped->nsenderbuffers];
				}

				int bufno= 0;

#if defined(DEBUG_TagOutgoingStreams)
				bufno++;
				overlapped->senderbuffers[0].buf = (char *) miniheader;
				overlapped->senderbuffers[0].len = 8;
				((int*) miniheader)[1] = seqno;
#endif

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt2 = clock->Time;
#endif

				for each (QS::Fx::Base::Block segment in request.Data.Segments)
				{
					switch (segment.type)
					{
					case QS::Fx::Base::Block::Type::Managed:
						{
							GCHandle bufferHandle = GCHandle::Alloc(segment.buffer, GCHandleType::Pinned);
							overlapped->senderbuffers[bufno].buf = 
								((char*) ((LPVOID) ((IntPtr) bufferHandle.AddrOfPinnedObject()))) + segment.offset;
							overlapped->senderbuffers[bufno].len = segment.size;
							bufno++;

							bufferHandles->Enqueue(bufferHandle);
						}
						break;

					case QS::Fx::Base::Block::Type::Unmanaged:
						{
							overlapped->senderbuffers[bufno].buf = ((char*) ((LPVOID) segment.address)) + segment.offset;
							overlapped->senderbuffers[bufno].len = segment.size;
							bufno++;
						}
						break;

					default:
						throw gcnew Exception("Unsupported type of memory block encountered.");
					}
				}

				data = request.Data;
				completionCallback = request.Callback;
				context = request.Context;

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt3 = clock->Time;
#endif

#if defined(DEBUG_DisplayOutgoingPackets)
				unsigned char _checksum = 0;
				System::Text::StringBuilder s;
				unsigned int _ppos = 0;
				for (unsigned int _ibuf = 0; _ibuf < nbuffers; _ibuf++)
				{
					WSABUF _buf = overlapped->senderbuffers[_ibuf];
					for (unsigned int _ichr = 0; _ichr < _buf.len; _ichr++)
					{
						unsigned char _c = _buf.buf[_ichr];
						_checksum ^= _c;
						if ((_ppos % 8) == 0)
							s.Append("\n" + _ppos.ToString("000000") + "-" + (_ppos + 7).ToString("000000"));
						s.Append(" ");
						s.Append(((unsigned int) _c).ToString("x2"));
						_ppos++;
					}
				}
				s.Append("\n\n");
				logger->Log("OUTGOING PACKET ( size = " + _ppos.ToString() + ", checksum = " + _checksum.ToString() + " ) : \n" + s.ToString());
#endif

				bool succeeded = false;
				DWORD numberOfBytes = 0;
				if (WSASend((SOCKET)((HANDLE) socket->Handle), overlapped->senderbuffers, nbuffers, 
					&numberOfBytes, 0,  (LPWSAOVERLAPPED) (&overlapped->overlapped), NULL) == SOCKET_ERROR)
				{
					int errorCode = WSAGetLastError();
					if (errorCode != WSA_IO_PENDING)
					{
						errorCallback("Could not initiate send operation, error: " + errorCode.ToString());
					}
					else
						succeeded = true;
				}
				else
				{
					succeeded = true;
					// errorCallback("Asynchronous send operation completed immediately, but it should not.");
				}

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt4 = clock->Time;
#endif

				if (!succeeded)
					Completed();

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt5 = clock->Time;

				cumulated_d1 += tt2 - tt1;
				cumulated_d2 += tt3 - tt2;
				cumulated_d3 += tt4 - tt3;
				cumulated_d4 += tt5 - tt4;
				nsamples++;

				if (tt5 > last_checked + 1)
				{
					owner->LogTransmitOverheads(this, tt1, cumulated_d1 / ((double) nsamples), cumulated_d2 / ((double) nsamples), 
						cumulated_d3 / ((double) nsamples), cumulated_d4 / ((double) nsamples));
					
					nsamples = 0;
					cumulated_d1 = cumulated_d2 = cumulated_d3 = cumulated_d4 = 0;
					last_checked = tt5;
				}
#endif
			}

			#pragma endregion

			#pragma region Completed

			void Transmitter::Completed()
			{
#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt1 = clock->Time;
#endif

				for each (GCHandle bufferHandle in bufferHandles)
					bufferHandle.Free();
				bufferHandles->Clear();

				int numberOfBytes = data.Size;
				QS::Fx::Base::ContextCallback^ callback = this->completionCallback;
				Object^ context = this->context;

				data.Clear();
				this->completionCallback = nullptr;
				this->context = nullptr;

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt2 = clock->Time;
#endif

				owner->RecycleTransmitter(this, numberOfBytes);

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionOverheads)
				double tt3 = clock->Time;
#endif

				if (callback != nullptr)
					callback(context);

#if defined(DEBUG_ProvideDetailedProfilingInformationForTransmissionCompletionOverheads)
				double tt4 = clock->Time;

				completion_cumulated_d1 += tt2 - tt1;
				completion_cumulated_d2 += tt3 - tt2;
				completion_cumulated_d3 += tt4 - tt3;
				completion_nsamples++;

				if (tt4 > completion_last_checked + 1)
				{
					owner->LogTransmitCompletionOverheads(this, tt1, 
						completion_cumulated_d1 / ((double) completion_nsamples), 
						completion_cumulated_d2 / ((double) completion_nsamples), 
						completion_cumulated_d3 / ((double) completion_nsamples));
					
					completion_nsamples = 0;
					completion_cumulated_d1 = completion_cumulated_d2 = completion_cumulated_d3 = 0;
					completion_last_checked = tt4;
				}
#endif
			}

			#pragma endregion
		}
	}
}
